lib/pipeline.js

```diff
@@ -1779,6 +1780,30 @@ class PipelineModel extends BaseModel {
             });
         };
 
+        const removeStagesAndStageBuilds = async () => {
+            // Lazy load factory dependency to prevent circular dependency issues
+            // https://nodejs.org/api/modules.html#modules_cycles
+            /* eslint-disable global-require */
+            const StageFactory = require('./stageFactory');
+            const StageBuildFactory = require('./stageBuildFactory');
+            /* eslint-enable global-require */
+
+            const stageFactory = StageFactory.getInstance();
+            const stageBuildFactory = StageBuildFactory.getInstance();
+
+            const stages = await stageFactory.list({ params: { pipelineId: this.id } });
+
+            logger.info(`pipelineId:${this.id}: Removing stages and stageBuilds.`);
+
+            stages.map(async s => {
+                const stageBuilds = await stageBuildFactory.list({ params: { stageId: s.id } });
+
+                await Promise.all(stageBuilds.map(sb => sb.remove()));
+            });
```

Comment:
We could potentially reduce the DB calls by getting stageBuilds for all the stages at once. 
```suggestion
            const stageIds = stages.map(s => s.id);
            const stageBuilds = await stageBuildFactory.list({ params: { stageId: stageIds} });
            await Promise.all(stageBuilds.map(sb => sb.remove()));
```

Reply:
updated